service/raft/raft_group0_client.cc

```diff
@@ -153,7 +153,9 @@ semaphore& raft_group0_client::operation_mutex() {
     return _operation_mutex;
 }
 
-future<> raft_group0_client::add_entry(group0_command group0_cmd, group0_guard guard, seastar::abort_source* as) {
+future<> raft_group0_client::add_entry(group0_command group0_cmd, group0_guard guard, seastar::abort_source* as,
+        std::optional<raft_timeout> timeout)
```

Comment:
I don't understand why you made these design choices.

Some callers of `raft_group0_client::add_entry` -- those that want to provide a `timeout` -- will provide both `as` and `timeout`.

So why not do the "composite abort_source" thing on the caller of `raft_group0_client::add_entry`? Why  introduce abstraction inside `raft_server_for_group`?

This way the interface of `raft_group0_client::add_entry` would still be simple (just taking `abort_source*`); if the caller wants timeout, they would pass a composite abort_source that includes a timeout source.

(etc. for other places)

Reply:
After reading the rest of the PR, it makes sense, especially with the extra reporting of quorum loss etc. (you'd basically have to wrap each raft::server API call with something that would do the reporting)